Written about how camf works (#34)
thesis20/report: sections/contextrecs/camf.tex

\subsection{Context-Aware Matrix Factorization}

> One of the context-aware recommender system methods we have looked into is context-aware matrix factorization (CAMF)\cite{baltrunasCAMF}.

Reviewer: ```suggestion
One approach to making a context-aware recommender system is \textit{context-aware matrix factorization} (CAMF)\cite{baltrunasCAMF}.
```

Author: Ved ikke lige om jeg er enig i den her

Reviewer: er mest af alt bare ikke fan af "we have looked into"

> As no implementation was available, we decided to make our own implementation of CAMF.

Reviewer: ```suggestion
As no implementation was available for the method, all results are based on our implementation.
```

> The method is an extention of the classical Matrix Factorization (MF) approach which takes the contextual side information into account when making rating predictions.

Reviewer: ```suggestion
The method is an extension of the classical Matrix Factorization (MF) approach which takes the contextual side information into account when making rating predictions.
```

> There are different models of CAMF that deal with different levels of granularity in terms of the context.

Reviewer: hvad vil det sige med granularity?

Author: hvordan context modelleres i forhold til interaktion med items. f.eks. at man enten siger at context har samme effektpå rating for alle items (CAMF-C) eller at man siger at context kan have forskellige effekt på rating alt efter hvilket item det er (CAMF-CI

> For CAMF-C, its assumed that each of the contextual conditions have a global influence on the ratings independently of the items.

Reviewer: ```suggestion
For CAMF-C, it is assumed that each of the contextual conditions has a global influence on the ratings independently of the items.
```

> In the CAMF-C model, a single parameter is introduced for each value of each contextual factor.

Reviewer: kan vi få lidt eksempel?

Author: Det kan vi godt men er det nødvendigt? Vi skal vel også tænke på hvor meget detalje vi vil gå eller hvor meget det skal fylde

The second model is CAMF-CI which introduces a parameter for each contextual condition and item pair, meaning that it has a finer granularity.

> By modelling it like this, it can capture when the contextual factor's have a different effect on the rating depending on which item it is.

Reviewer: ```suggestion
By modeling it like this, it is possible to capture when the contextual factors have a different effect on the rating depending on which item it is.
```

er ikke sikker på hvad den her sætning faktisk betyder

> The expression for predicting ratings with the CAMF models is seen in \autoref{eqn:camf-rating-pred}.

Reviewer: er expression det rigtige ord? @Eduardsen

\begin{equation}
    \label{eqn:camf-rating-pred}
    \hat{r}_{uic_1...c_k} = \vec{v_u} * \vec{q_i} + \bar{i} + b_u + \sum\limits_{j = 1}^k B_{ijc_j}
\end{equation}
We have that $\hat{r}_{uic_1...c_k}$ is the predicted rating for user $u$, item $i$ under the contextual values $c_1...c_k$.

> The variables $\vec{v_u} $ and $ \vec{q_i}$ are the feature vectors for user $u$ and item $i$ which are multiplied, just as in matrix factorization.

Reviewer: vi skal nok enten beskrive matrix factorization, eller fortælle hvad en feature vector er her

Author: Tror du ikke godt vi kan forvente at læseren ved det? Ved ikke om vejlderne har sagt noget omkring det men hvis man sammenligner med papers man læser går de vel heller ikke meget i detaljer om det. Men man kan vel godt lige hurtigt beskrive det hvis det er nødvendigt

> $\bar{i}$ is the global average rating for item $i$ and $\bar{b_u}$ is a user bias. 

Reviewer: ```suggestion
$\bar{i}$ is the global average rating for item $i$, meaning the average rating across all users in the training set, and $\bar{b_u}$ is a user bias.
```

$B_{ijc_j}$ are the parameters modeling the interaction of the contextual conditions and the items.

> For the CAMF-CI model, $B_{ijc_j}$ will result a lot of parameters, and for CAMF-C it will result in less, as a results of how the two differnet approaches model the context.

Reviewer: ```suggestion
For the CAMF-CI model, $B_{ijc_j}$ will result in a lot of parameters, and for CAMF-C it will result in less, as a result of how the two different approaches perform the contextual modeling.
```

> All of the parameters in the model, except for the average item rating, are learnt through stochastic gradient descent.

Reviewer: ```suggestion
All of the parameters in the model, except for the average item rating, are learned through stochastic gradient descent.
```

This is done through the use of a loss function which is seen in \autoref{eqn:camf-loss-func}.
\begin{equation}
    \label{eqn:camf-loss-func}
    \begin{split}
        \min_{v*, q*, b*, B*}\sum \limits_{r \in  R}\left [ \left (  \hat{r}_{uic_1...c_k} - \vec{v_u} * \vec{q_i} - \bar{i} - b_u - \sum\limits_{j = 1}^k B_{ijc_j}\right )^2 \right. \\
        \left. + \lambda \left({b_u}^2 +{\left \| \vec{v_u} \right \|}^2  + {\left \|\vec{q_i}  \right \|}^2 + \sum\limits_{j = 1}^k \sum\limits_{c_j = 1}^{z_j} B_{ijc_j}^{2}\right ) \right ]
    \end{split}
\end{equation}
Here we have that $r = (u,i,c_1...c_k)$ and that R is the context-dependent ratings from the training set.
The loss function includes regularization controlled by the $\lambda$ parameter to avoid overfitting the training data.

> The models parameters are then updated based on the gradient of the loss function for each of the ratings in the training set.

Reviewer: ```suggestion
The parameters are then updated based on the gradient of the loss function for each of the ratings in the training set.
```

\subsection{Context-Aware Matrix Factorization}
One approach to making a context-aware recommender system is \textit{context-aware matrix factorization} (CAMF)\cite{baltrunasCAMF}.
As no implementation was available for the method, all results are based on our implementation.
The method is an extension of the classical Matrix Factorization (MF) approach which takes the contextual side information into account when making rating predictions.
There are different models of CAMF that deal with different levels of granularity in terms of the context.
For CAMF-C, it is assumed that each of the contextual conditions has a global influence on the ratings independently of the items.
In the CAMF-C model, a single parameter is introduced for each value of each contextual factor.
The second model is CAMF-CI which introduces a parameter for each contextual condition and item pair, meaning that it has a finer granularity.
By modelling it like this, it can capture when the contextual factor's have a different effect on the rating depending on which item it is.
In this case, CAMF is able to capture two different levels of granularity.

> With CAMF-C for example, we have that when it is raining outside as the contextual value, every movie will get an increased predicted rating across the board.

Reviewer 2: ```suggestion
With CAMF-C, for example, every movie will get an increased predicted rating across the board when it is raining outside as the contextual value.
```

> For CAMF-CI however, we have that when it is raining outside as the contextual value, only some specific movies will have their predicted ratings increased, while others may have their predicted ratings decreased.

Reviewer 2: ```suggestion
For CAMF-CI, however, only some specific movies will have their predicted ratings increased, while others may have their predicted ratings decreased when raining outside is used as the contextual value.
```

The expression for predicting ratings with the CAMF models is seen in \autoref{eqn:camf-rating-pred}.
\begin{equation}
    \label{eqn:camf-rating-pred}
    \hat{r}_{uic_1...c_k} = \vec{v_u} * \vec{q_i} + \bar{i} + b_u + \sum\limits_{j = 1}^k B_{ijc_j}
\end{equation}

> We have that $\hat{r}_{uic_1...c_k}$ is the predicted rating for user $u$, item $i$ under the contextual values $c_1...c_k$.

Reviewer 2: ```suggestion
We have that $\hat{r}_{uic_1...c_k}$ is the predicted rating for user $u$ for item $i$ under the contextual values $c_1...c_k$.
```

For the CAMF-CI model, $B_{ijc_j}$ will result in a lot of parameters, and for CAMF-C it will result in less, as a result of how the two different approaches perform the contextual modeling.
All of the parameters in the model, except for the average item rating, are learned through stochastic gradient descent.
This is done through the use of a loss function which is seen in \autoref{eqn:camf-loss-func}.
\begin{equation}
    \label{eqn:camf-loss-func}
    \begin{split}
        \min_{v*, q*, b*, B*}\sum \limits_{r \in  R}\left [ \left (  \hat{r}_{uic_1...c_k} - \vec{v_u} * \vec{q_i} - \bar{i} - b_u - \sum\limits_{j = 1}^k B_{ijc_j}\right )^2 \right. \\
        \left. + \lambda \left({b_u}^2 +{\left \| \vec{v_u} \right \|}^2  + {\left \|\vec{q_i}  \right \|}^2 + \sum\limits_{j = 1}^k \sum\limits_{c_j = 1}^{z_j} B_{ijc_j}^{2}\right ) \right ]
    \end{split}
\end{equation}
Here we have that $r = (u,i,c_1...c_k)$ and that R is the context-dependent ratings from the training set.
The loss function includes regularization controlled by the $\lambda$ parameter to avoid overfitting the training data.

> The parameters are then updated based on the gradient of the loss function for each of the ratings in the training set.

Reviewer 2: hvad er z og k i loss funktionen?

Author: fixed i senest commit
